Tests/PowerSync/PowerSync.Common.Tests/Client/PowerSyncDatabaseTransactionTests.cs

```diff
@@ -0,0 +1,378 @@
+    }
+
+    [Fact]
+    public async Task WriteTransactionNestedQueryTest()
+    {
+        await db.WriteTransaction(async tx =>
+        {
+            await tx.Execute("INSERT INTO assets(id) VALUES(?)", ["O6"]);
+
+            var txQuery = await tx.GetAll<IdResult>("SELECT * FROM assets");
+            Assert.Single(txQuery);
+
+            var dbQuery = await db.GetAll<IdResult>("SELECT * FROM assets");
+            Assert.Empty(dbQuery);
+        });
+    }
+
+    [Fact]
+    public async Task ReadLockReadOnlyTest()
+    {
+        string id = Guid.NewGuid().ToString();
+        bool exceptionThrown = false;
+
+        try
+        {
+            await db.ReadLock<object>(async context =>
+            {
+                return await context.Execute(
+                    "INSERT INTO assets (id) VALUES (?)",
+                    [id]
+                );
+            });
+
+            // If no exception is thrown, fail the test
+            throw new Exception("Did not throw");
+        }
+        catch (Exception ex)
+        {
+            Assert.Contains("attempt to write a readonly database", ex.Message);
+            exceptionThrown = true;
+        }
+
+        Assert.True(exceptionThrown);
+    }
+
+    [Fact]
+    public async Task ReadLocksQueueIfExceedNumberOfConnectionsTest()
+    {
+        string id = Guid.NewGuid().ToString();
+
+        await db.Execute(
+            "INSERT INTO assets (id) VALUES (?)",
+            [id]
+        );
+
+        int numberOfReads = 20;
+        var tasks = Enumerable.Range(0, numberOfReads)
+            .Select(_ => db.ReadLock(async context =>
+            {
+                return await context.GetAll<AssetResult>("SELECT id FROM assets WHERE id = ?", [id]);
+            }))
+            .ToArray();
+
+        var lockResults = await Task.WhenAll(tasks);
+
+        var ids = lockResults.Select(r => r.FirstOrDefault()?.id).ToList();
+
+        Assert.All(ids, n => Assert.Equal(id, n));
+    }
+
+    [Fact(Timeout = 2000)]
+    public async Task ReadWhileWriteIsRunningTest()
+    {
+        var tcs = new TaskCompletionSource();
+
+        // This wont resolve or free until another connection free's it
+        var writeTask = db.WriteLock(async context =>
+        {
+            await tcs.Task; // Wait until read lock signals to proceed
+        });
+
+        var readTask = db.ReadLock(async context =>
+        {
+            // Read logic could execute here while writeLock is still open
+            tcs.SetResult();
+            await Task.CompletedTask;
+            return 42;
+        });
+
+        var result = await readTask;
+        await writeTask;
+
+        Assert.Equal(42, result);
+    }
+
+    [Fact(Timeout = 2000)]
+    public async Task QueueSimultaneousExecutionsTest()
+    {
+        var order = new List<int>();
+        var operationCount = 5;
+
+        await db.WriteLock(async context =>
+        {
+            var tasks = Enumerable.Range(0, operationCount)
```

Comment:
Another thing to test would be the case of not awaiting all context operations inside this callback. I.e. using the `context` outside of the lambda should not be allowed.